### PR TITLE
ADL terminology generator

### PR DESCRIPTION
**Pull request by pieterbos**
A tool to make ADL editing by hand easier:

Simply write your ADL, including comments after the node ids with the text. Add the languages you want to the terminology, with no or just a few entries per language. Then run it through the adlchecker with the '--lint' option.

Now you'll receive a correctly indented version, with all missing terms in the terminology added automatically.

at- and ac-codes get added with a text marking they should be translated. id codes with a comment on the same line where it is used in the definition get that comment added to both its text and definition, so you can easily complete the terminology. If no comment has been provided it gets added with a 'translate me' text+description. If the id code translation is not mandatory in the terminology (DV_* in ELEMENT.value, for example), it does not get added at all.

Saves a lot of time if you don't have a functional ADL 2 compatible archetype editor...

### CODE REVIEW

File: adlchecker/src/main/java/com/nedap/archie/adlchecker/TerminologyContentGenerator.java [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+        }
+        return null;
+    }
+
+    public boolean terminologyHasCode(Archetype archetype, CObject cObject) {
+        String nodeId = cObject.getNodeId();
+        int codeSpecializationDepth = AOMUtils.getSpecializationDepthFromCode(nodeId);
+        int archetypeSpecializationDepth = archetype.specializationDepth();
+        if(codeSpecializationDepth > archetypeSpecializationDepth) {
+            return true;//not exactly, this is a validation failure that needs to be fixed. log?
```
Comment (t-de-jong) [inline adlchecker/src/main/java/com/nedap/archie/adlchecker/TerminologyContentGenerator.java]: Can this be fixed now? Or should a new ticket be created for this?

File: adlchecker/src/main/java/com/nedap/archie/adlchecker/TerminologyContentGenerator.java [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+            sortedArchetypeTermMap.putAll(stringArchetypeTermMap);
+            termDefinitions.put(language, sortedArchetypeTermMap);
+
+        }
+    }
+
+    private void walkArchetype(String sourceFile, Archetype archetype, Archetype resultingArchetype) {
+        Stack<CObject> workList = new Stack<>();
+        workList.push(archetype.getDefinition());
+        List<CObject> cObjects = new ArrayList<>();
```
Comment (t-de-jong) [inline adlchecker/src/main/java/com/nedap/archie/adlchecker/TerminologyContentGenerator.java]: When is cObjects actually used?

File: adlchecker/src/main/java/com/nedap/archie/adlchecker/TerminologyContentGenerator.java [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+                newCode.setText(text);
+                newCode.setDescription(description);
+                termDefinitions.get(language).put(next.getNodeId(), newCode);
+            }
+        }
+    }
+
+    private String getCommentName(String sourceFile, String nodeId) {
+        String[] lines = sourceFile.split("\n");
+        for(String line:lines) {
```
Comment (t-de-jong) [inline adlchecker/src/main/java/com/nedap/archie/adlchecker/TerminologyContentGenerator.java]: Does this retrieve all the comments in the archetype every time when you need one specific comment?

File: adlchecker/src/main/java/com/nedap/archie/adlchecker/TerminologyContentGenerator.java [hunk 4]
```diff
@@ -140,6 +170,28 @@ public boolean terminologyHasCode(Archetype archetype, CObject cObject) {
...
+     * Check if terminology has a code. Use only for at- and ac-codes, NOT For id codes!!
+     * @param archetype
+     * @param code
+     * @return
+     */
+    public boolean terminologyHasCodeForAllLanguages(Archetype archetype, String code) {
+        if(AOMUtils.isIdCode(code)) {
+            throw new IllegalArgumentException("this method only checks at- and ac- codes, not id codes");
+        }
+        int codeSpecializationDepth = AOMUtils.getSpecializationDepthFromCode(code);
```
Comment (t-de-jong) [inline adlchecker/src/main/java/com/nedap/archie/adlchecker/TerminologyContentGenerator.java]: This method mostly contains duplicate code from terminologyHasCodeForAllLanguages. Maybe extract this code or something?

File: adlchecker/src/main/java/com/nedap/archie/adlchecker/TerminologyContentGenerator.java [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+            sortedArchetypeTermMap.putAll(stringArchetypeTermMap);
+            termDefinitions.put(language, sortedArchetypeTermMap);
+
+        }
+    }
+
+    private void walkArchetype(String sourceFile, Archetype archetype, Archetype resultingArchetype) {
+        LinkedList<CObject> workList = new LinkedList<>();
+        workList.push(archetype.getDefinition());
+        List<CObject> cObjects = new ArrayList<>();
```
Comment (t-de-jong) [inline adlchecker/src/main/java/com/nedap/archie/adlchecker/TerminologyContentGenerator.java]: When is cObjects actually used?

File: adlchecker/src/main/java/com/nedap/archie/adlchecker/TerminologyContentGenerator.java [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+            }
+        }
+    }
+
+    private void addCodeToTerminology(String sourceFile, Archetype resultingArchetype, String code) {
+        String text = getCommentName(sourceFile, code);
+        if(text == null) {
+            text = "Add term for me!";
+        }
+        String description = text; //TODO: grep comment from sourcefile!
```
Comment (t-de-jong) [inline adlchecker/src/main/java/com/nedap/archie/adlchecker/TerminologyContentGenerator.java]: Isn't the todo done by getCommentName()?

File: adlchecker/src/main/java/com/nedap/archie/adlchecker/AdlChecker.java [hunk 7]
```diff
@@ -78,6 +88,38 @@ private static void validateArchetypes(List<String> directories, boolean printFl
             }
 
         }
+
+        if(lint) {
+            System.out.println("step 1: running archetypes through linter");
```
[the remark this answers is not in the document]
Comment (pieterbos) [reply adlchecker/src/main/java/com/nedap/archie/adlchecker/AdlChecker.java]: fixed in new branch

File: adlchecker/src/main/java/com/nedap/archie/adlchecker/TerminologyContentGenerator.java (hunk 5 above)
Comment (pieterbos) [reply adlchecker/src/main/java/com/nedap/archie/adlchecker/TerminologyContentGenerator.java]: removed in next branch

File: adlchecker/src/main/java/com/nedap/archie/adlchecker/TerminologyContentGenerator.java (hunk 6 above)
Comment (pieterbos) [reply adlchecker/src/main/java/com/nedap/archie/adlchecker/TerminologyContentGenerator.java]: fixed in next branch

File: adlchecker/src/main/java/com/nedap/archie/adlchecker/TerminologyContentGenerator.java (hunk 3 above)
Comment (pieterbos) [reply adlchecker/src/main/java/com/nedap/archie/adlchecker/TerminologyContentGenerator.java]: yes it does. Works fine for now, I'll add a comment

File: adlchecker/src/main/java/com/nedap/archie/adlchecker/TerminologyContentGenerator.java (hunk 1 above)
Comment (pieterbos) [reply adlchecker/src/main/java/com/nedap/archie/adlchecker/TerminologyContentGenerator.java]: changed comment with explanation - this cannot be fixed, must remain as is and will be logged by the archetype validator.

File: adlchecker/src/main/java/com/nedap/archie/adlchecker/TerminologyContentGenerator.java (hunk 4 above)
Comment (pieterbos) [reply adlchecker/src/main/java/com/nedap/archie/adlchecker/TerminologyContentGenerator.java]: except that it's not and it does something different and extracting the code would make it much harder to read

**Comment by pieterbos** [comment]
Continued in https://github.com/openEHR/archie/pull/31, review comments that resulted in code changes have been done in that branch

**Pull request closed without merging**